### ISSUE TITLE
Typetracer arrays: support for more operations based on dask-awkward's necessary columns optimization

### ISSUE THREAD

**Issue by douglasdavis** [scikit-hep/awkward#2135]
The new typetracer report optimization PR in dask-awkward (https://github.com/dask-contrib/dask-awkward/pull/141) is surfacing some features that are missing for typetracers. Creating this issue to track some findings from the dask-awkward test suite that have surfaced after implementing an optimization using the new typetracer data-touch feature.

It's definitely possible some of these may be issues in dask-awkward itself, so please let us know if you think that's the case.

- [ ] A slicing error:
	```
	  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/recordarray.py", line 300, in _getitem_at
	    raise ak._errors.index_error(self, where)
	IndexError: while attempting to slice
	
	    <Array-typetracer [...] type='?? * {run: int64, luminosityBlock: int64,...'>
	
	with
	
	    0
	
	at inner RecordArray of length 0, using sub-slice 0.
	```
- [x] A truedivide `NotImplementedError`
- [x] `nplikes.Numpy.asarray` not compatible with `UnknownScalar`
	```
	    def asarray(self, *args, **kwargs):
	        # array[, dtype=][, order=]
	>       return self._module.asarray(*args, **kwargs)
	E       TypeError: int() argument must be a string, a bytes-like object or a real number, not 'UnknownScalar'
	```
- [ ] Error with building an array of UnknownScalars. In `ak_from_iter`:
	```
	        builder = _ext.ArrayBuilder(initial=initial, resize=resize)
	>       builder.fromiter(iterable)
	E       ValueError: cannot convert MaybeNone(UnknownScalar(dtype('int64'))) (type MaybeNone) to an array element
	E       
	E       (https://github.com/scikit-hep/awkward/blob/awkward-cpp-7/awkward-cpp/src/python/content.cpp#L191)
	
	allow_record = True
	behavior   = None
	builder    = <awkward_cpp.lib._ext.ArrayBuilder object at 0x148e4ff30>
	highlevel  = True
	initial    = 1024
	iterable   = [MaybeNone(UnknownScalar(dtype('int64'))), MaybeNone(UnknownScalar(dtype('int64'))), MaybeNone(UnknownScalar(dtype('int64')))]
	resize     = 1.5
	```

To actually run the test suite one can check out PR https://github.com/dask-contrib/dask-awkward/pull/141 

Cc: @martindurant

**Comment by agoose77** [comment]
Most of these will be fixed with work that I'm currently undertaking. I'm trying to merge it in piecemeal to keep the PR overhead down, but I'll start testing against the dask test-suite to ensure that we cover all of these cases.

**Comment by douglasdavis** [comment]
Angus Hollands ***@***.***> writes:

> Most of these will be fixed with work that I'm currently undertaking. I'm trying to merge it in piecemeal to keep the PR overhead down, but I'll start testing against the dask test-suite to ensure that we cover all of these cases.

Great! Like I said above, I think there's a good chance we have some
imperfect stuff in dask-awkward itself, so if something pops up/remains
an error in the test suite and it's not obviously originating from
awkward- then there is a possibility of a bug in the dask-awkward
optimization code

**Comment by martindurant** [comment]
> there is a possibility of a bug in the dask-awkward
optimization code

Almost certain :)

**Comment by douglasdavis** [comment]
Here's more detail on the array building error now that we don't have UnknownScalar anymore (I was writing up another issue but then realized this already came up before the `UnknownScalar --> TypeTracerArray(shape=())` transition!)

In dask-awkward reducers we take advantage of a generic tree reduction algorithm that is already part of upstream dask. (more details: it's the [`DataFrameTreeReduction`](https://github.com/dask/dask/blob/c6b7052acc52b25facb76e8b128d574db8b0fc9f/dask/layers.py#L1253) layer, which even though it has DataFrame in the name it doesn't depend on pandas, it was just originally designed for dask.dataframe).

This algorithm includes a concatenation step for combining intermediate results during the reduction.

For total reductions in dask-awkward, we use `ak.from_iter`.

For example, let's take `dak.sum(..., axis=None)`. If we have a partitioned array and we take the sum of each partition

```
x = dak.Array([[1,2,3], [4],      [5,6], [7],         [8,9], [10,11],        [12,13,14]])
               ^ partition 0      ^ partition 1       ^ partition 2          ^ partition 3

sum                10                    18                38                39
```

If the tree reduction settings are set to combine 2 partitions at a time, the next step will use `ak.from_iter` to create a new intermediate step like so:

```
     [[10, 18]              [38, 39]]
      ^ partition 0        ^ partition 1

sum       28                  77
```

Then another `ak.from_iter` to combine `[28, 77]` and finally take the sum of that to get the result `105`

When executing the typetracer-only dask graph the scalars `28` and `77` for example are passed to `ak.from_iter` as:

```python
ak.from_iter([TypeTracerArray(dtype('int64'), shape=()), TypeTracerArray(dtype('int64'), shape=()])
```

And this fails with the following traceback

<details>
<summary>click to reveal</summary>

```
    def _impl(iterable, highlevel, behavior, allow_record, initial, resize):
        if isinstance(iterable, dict):
            if allow_record:
                return _impl(
                    [iterable],
                    highlevel,
                    behavior,
                    False,
                    initial,
                    resize,
                )[0]
            else:
                raise ak._errors.wrap_error(
                    ValueError(
                        "cannot produce an array from a single dict (that would be a record)"
                    )
                )
    
        if isinstance(iterable, tuple):
            iterable = list(iterable)
    
        builder = _ext.ArrayBuilder(initial=initial, resize=resize)
>       builder.fromiter(iterable)
E       ValueError: cannot convert TypeTracerArray(dtype('int64'), shape=()) (type TypeTracerArray) to an array element
E       
E       (https://github.com/scikit-hep/awkward/blob/awkward-cpp-7/awkward-cpp/src/python/content.cpp#L191)
```
</details>

I'm thinking of 3 paths here:
1. feature request for support scalar TypeTracerArray elements in `ak.from_iter`
2. Someone tells me something else I can use other than from_iter :)
3. We create a code path in dask-awkward reducers to check for these scalar `TypeTracerArray` objects _before_ calling `from_iter`, and instead create a new typetracer array with the same dtype as the scalar to represent the intermediate step in typetracer form. (we already have this "special cases" treatment around typetracers with input and output layers, but adding more special cases for regular array operations seems like it could become fragile, but maybe there would only be a few of these)

**Comment by douglasdavis** [comment]
Here's a commit that implements a workaround: https://github.com/dask-contrib/dask-awkward/pull/148/commits/b74d912878e6009c9c7183789b3c619893e73af0

**Comment by agoose77** [comment]
You're right that this is something that we need to think about. In fact, the reason I started this work on nplikes was to fix problems with partial reductions for `axis=None`, which is very similar to this one. We ended up being able to take the best approach of removing `flatten_records`, such that no branching could take place during reduction. This meant we only ever had one contiguous array to reduce. 

However, in this case one does not have such a luxury. As I see it, the benefit to building an intermediate Awkward Array is that it should understand the option types that `reduce` returns "for free". Yet, we can't use `from_iter` (at least, not as it stands) for the reasons that you give; it's just not aware of typetracer types.
Furthermore, one cannot apply this two-stage reduction (at least not trivially) to positional reducers like `argmax`.

In any case,`from_iter` isn't ideal because it loses the type information of the scalar objects that we return, both for typetracer arrays and regular NumPy arrays. Instead, if we *know* that we're getting typetracer scalars (or None), then I think we want build the intermediate length-1 arrays before concatenating them.

AAUI, you only need this in `axis=None` reductions, right? The solution that we were pursuing is to completely flatten the array, reduce those arrays at `axis=-1` to preserve the array structure, and then use `concatenate` to build the final array that is reduced. You'll still need special handling for positional reducers, but it's possible to express this using exclusively L1 and L2 content manipulation, so that you don't need to get too much into the weeds of `MaybeNone` etc.

I'd suggest something like
```python
def primary_reduction(array, reducer):
    flat = ak.ravel(array)
    return reducer(flat, keepdims=True, axis=-1)

partials = [primary_reduction(x, reducer) for x in arrays]
array = ak.concatenate(partials, axis=0)

result = secondary_reducer(array, axis=None)
```

This would need care w.r.t positional reducers, but should avoid the need to handle `None` and `MaybeNone`, etc.
